### PR TITLE
Implement the Data Inspector

### PR DESCRIPTION
**Pull request by dak98**
This PR contains code additions necessary to implement the Data Inspector functionality.

Links:
- Data Inspector final presentation: https://indico.cern.ch/event/1004294/contributions/4217207/attachments/2185854/3693256/data-inspector-08-02-2021.pdf
- Plugin and proxy: https://github.com/dak98/DataInspector
- Client GUI: https://github.com/dak98/DataInspectorGUI

### CODE REVIEW

File: dependencies/O2Dependencies.cmake [hunk 1]
```diff
@@ -48,6 +48,13 @@ set_package_properties(ROOT PROPERTIES TYPE REQUIRED)
 find_package(fmt)
 set_package_properties(fmt PROPERTIES TYPE REQUIRED)
 
+find_package(PkgConfig)
+set_package_properties(PkgConfig PROPERTIES TYPE REQUIRED)
+
+# TODO: Pass this path as an -D cmake argument, avoiding `ubuntu2004_x86-64`
+set(ENV{PKG_CONFIG_PATH} "$ENV{ALIBUILD_WORK_DIR}/ubuntu2004_x86-64/ZeroMQ/latest/lib/pkgconfig")
```
Comment (ktf) [inline dependencies/O2Dependencies.cmake:55]: This clearly needs to go.

File: Framework/Core/CMakeLists.txt [hunk 2]
```diff
@@ -127,6 +128,7 @@ o2_add_library(Framework
                                      ${BOOST_FILESYSTEM}
                                      arrow::gandiva_shared
                                      LibUV::LibUV
+                                     PkgConfig::ZeroMQ
```
Comment (ktf) [inline Framework/Core/CMakeLists.txt:135]: Why is this needed at all?

File: Framework/Core/src/DataInspector.cxx [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+namespace o2
+{
+namespace framework
+{
+
+bool isDataInspectorActive(FairMQDevice& device)
+try {
+  const char* value = std::getenv("DEVICES");
```
Comment (ktf) [inline Framework/Core/src/DataInspector.cxx]: I do not think getting the list from an environment variable is a good idea. Can't you simply provide it as an argument to the `--inspector` option?

File: Framework/Core/src/DataInspector.cxx [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    pos = data.find(pattern, pos + substitute.size());
+  }
+}
+
+namespace o2
+{
+namespace framework
+{
+
+bool isDataInspectorActive(FairMQDevice& device)
```
Comment (ktf) [inline Framework/Core/src/DataInspector.cxx:201]: Do I understand correctly that you need to do a string comparison, a getenv and inspect the options for each message, in order to decide whether it should be copied and inspected?

File: Framework/Core/CMakeLists.txt (hunk 2 above)
Comment (dak98) [reply Framework/Core/CMakeLists.txt:135]: DataInspector files use the ZeroMQ library.

File: dependencies/O2Dependencies.cmake (hunk 1 above)
Comment (dak98) [reply dependencies/O2Dependencies.cmake:55]: I agree but I honestly don't know how to do it (I put some time into it but failed). That's why I put this comment.

File: Framework/Core/src/DataInspector.cxx (hunk 4 above)
Comment (dak98) [reply Framework/Core/src/DataInspector.cxx:201]: - I use getenv to set which devices should to be inspected from the start.
- I use ProgOptions to set which devices should be inspected dynamically.

File: Framework/Core/src/DataInspector.cxx [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+try {
+  if (device.fConfig == nullptr) {
+    return false;
+  }
+  return device.fConfig->GetProperty<bool>(INSPECTOR_ACTIVATION_PROPERTY);
```
Comment (ktf) [inline Framework/Core/src/DataInspector.cxx:206]: This still does a string lookup per message. Can it please be reworked so that is not needed?

File: dependencies/O2Dependencies.cmake (hunk 1 above)
Comment (ktf) [reply dependencies/O2Dependencies.cmake:55]: Ok, but this needs to be cleaned in any case. Why not simply using FairMQ, BTW?

File: Framework/Core/src/DataInspector.cxx (hunk 5 above)
Comment (dak98) [reply Framework/Core/src/DataInspector.cxx:206]: I don't think it can be reworked. This mechanism is used to dynamically enable or disable the sending of messages to the custom, DataInspector device. It means, that the value of the lookup can change between the checks. I also benchmarked the DataInspector using qc and I didn't find any performance loss.

File: dependencies/O2Dependencies.cmake (hunk 1 above)
Comment (dak98) [reply dependencies/O2Dependencies.cmake:55]: The ZeroMQ library is used to move the inspected messages out of the O2 analysis completely and into a separate component. I'm not sure whether it is possible to implement this using FairMQ. Plus, this separate component uses ZeroMQ for other purposes as well.

File: Framework/Core/src/DataInspector.cxx (hunk 4 above)
Comment (jotwinow) [reply Framework/Core/src/DataInspector.cxx:201]: @ktf @dak98 and will you find time to go through this PR.
